File: src/test/java/org/apache/avro/TestReflect.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.avro;

import static org.junit.Assert.assertEquals;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Collection;
import java.util.Map;
import java.util.List;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Arrays;

import org.apache.avro.TestReflect.SampleRecord.AnotherSampleRecord;
import org.apache.avro.io.BinaryDecoder;
import org.apache.avro.io.BinaryEncoder;
import org.apache.avro.reflect.ReflectData;
import org.apache.avro.reflect.ReflectDatumReader;
import org.apache.avro.reflect.ReflectDatumWriter;

import org.junit.Test;

public class TestReflect {

  @Test public void testVoid() {
    check(Void.TYPE, "\"null\"");
    check(Void.class, "\"null\"");
  }

  @Test public void testBoolean() {
    check(Boolean.TYPE, "\"boolean\"");
    check(Boolean.class, "\"boolean\"");
  }

  @Test public void testInt() {
    check(Integer.TYPE, "\"int\"");
    check(Integer.class, "\"int\"");
  }

  @Test public void testLong() {
    check(Long.TYPE, "\"long\"");
    check(Long.class, "\"long\"");
  }

  @Test public void testFloat() {
    check(Float.TYPE, "\"float\"");
    check(Float.class, "\"float\"");
  }

  @Test public void testDouble() {
    check(Double.TYPE, "\"double\"");
    check(Double.class, "\"double\"");
  }

  @Test public void testString() {
    check("Foo", "\"string\"");
  }

  @Test public void testBytes() {
    check(new byte[0], "\"bytes\"");
  }

  public static class R1 {
    private Map<String,String> mapField = new HashMap<String,String>();
    private String[] arrayField = new String[] { "foo" };
    private List<String> listField = new ArrayList<String>();

    {
      mapField.put("foo", "bar");
      listField.add("foo");
    }
    
    public boolean equals(Object o) {
      if (!(o instanceof R1)) return false;
      R1 that = (R1)o;
      return mapField.equals(that.mapField)
        && Arrays.equals(this.arrayField, that.arrayField) 
        &&  listField.equals(that.listField);
    }
  }

  @Test public void testMap() throws Exception {
    check(R1.class.getDeclaredField("mapField").getGenericType(),
          "{\"type\":\"map\",\"values\":\"string\"}");
  }

  @Test public void testArray() throws Exception {
    check(R1.class.getDeclaredField("arrayField").getGenericType(),
          "{\"type\":\"array\",\"items\":\"string\"}");
  }
  @Test public void testList() throws Exception {
    check(R1.class.getDeclaredField("listField").getGenericType(),
          "{\"type\":\"array\",\"items\":\"string\"}");
  }

  @Test public void testR1() throws Exception {
    checkReadWrite(new R1());
  }

  public static class R2 {
    private String[] arrayField;
    private Collection<String> collectionField;
    
    public boolean equals(Object o) {
      if (!(o instanceof R2)) return false;
      R2 that = (R2)o;
      return Arrays.equals(this.arrayField, that.arrayField) 
        &&  collectionField.equals(that.collectionField);
    }
  }

  @Test public void testR2() throws Exception {
    R2 r2 = new R2();
    r2.arrayField = new String[] {"foo"};
    r2.collectionField = new ArrayList<String>();
    r2.collectionField.add("foo");
    checkReadWrite(r2);
  }

  public static class R3 {
    private int[] intArray;
    
    public boolean equals(Object o) {
      if (!(o instanceof R3)) return false;
      R3 that = (R3)o;
      return Arrays.equals(this.intArray, that.intArray);
    }
  }

  @Test public void testR3() throws Exception {
    R3 r3 = new R3();
    r3.intArray = new int[] {1};
    checkReadWrite(r3);
  }

  public static class R4 {
    public short value;
    
    public boolean equals(Object o) {
      if (!(o instanceof R4)) return false;
      return this.value == ((R4)o).value;
    }
  }

  public static class R5 extends R4 {}

  @Test public void testR5() throws Exception {
    R5 r5 = new R5();
    r5.value = 1;
    checkReadWrite(r5);
  }

  void checkReadWrite(Object object) throws Exception {
    Schema s = ReflectData.get().getSchema(object.getClass());
    ReflectDatumWriter writer = new ReflectDatumWriter(s);
    ByteArrayOutputStream out = new ByteArrayOutputStream();
    writer.write(object, new BinaryEncoder(out));
    ReflectDatumReader reader = new ReflectDatumReader(s);
    Object after =
      reader.read(null, new BinaryDecoder
                  (new ByteArrayInputStream(out.toByteArray())));
    assertEquals(object, after);
  }

  public static enum E { A, B };
  @Test public void testEnum() throws Exception {
    check(E.class, "{\"type\":\"enum\",\"name\":\"E\",\"namespace\":"
          +"\"org.apache.avro.TestReflect$\",\"symbols\":[\"A\",\"B\"]}");
  }

  public static class R { int a; long b; }
  @Test public void testRecord() throws Exception {
    check(R.class, "{\"type\":\"record\",\"name\":\"R\",\"namespace\":"
          +"\"org.apache.avro.TestReflect$\",\"fields\":["
          +"{\"name\":\"a\",\"type\":\"int\"},"
          +"{\"name\":\"b\",\"type\":\"long\"}]}");
  }

  private void check(Object o, String schemaJson) {
    check(o.getClass(), schemaJson);
  }

  private void check(Type type, String schemaJson) {
    assertEquals(schemaJson, ReflectData.get().getSchema(type).toString());
  }

  @Test
  public void testRecordIO() throws IOException {
    Schema schm = ReflectData.get().getSchema(SampleRecord.class);
    ReflectDatumWriter writer = new ReflectDatumWriter(schm);
    ByteArrayOutputStream out = new ByteArrayOutputStream();
    SampleRecord record = new SampleRecord();
    record.x = 5;
    record.y = 10;
    writer.write(record, new BinaryEncoder(out));
    ReflectDatumReader reader = new ReflectDatumReader(schm);
    Object decoded =
      reader.read(null, new BinaryDecoder
                  (new ByteArrayInputStream(out.toByteArray())));
    assertEquals(record, decoded);
  }

  @Test
  public void testRecordWithNullIO() throws IOException {
    ReflectData reflectData = ReflectData.AllowNull.get();
    Schema schm = reflectData.getSchema(AnotherSampleRecord.class);
    ReflectDatumWriter writer = new ReflectDatumWriter(schm);
    ByteArrayOutputStream out = new ByteArrayOutputStream();
    // keep record.a null and see if that works
    AnotherSampleRecord a = new AnotherSampleRecord();
    writer.write(a, new BinaryEncoder(out));
    AnotherSampleRecord b = new AnotherSampleRecord(10);
    writer.write(b, new BinaryEncoder(out));
    ReflectDatumReader reader = new ReflectDatumReader(schm);
    ByteArrayInputStream in = new ByteArrayInputStream(out.toByteArray());
    Object decoded = reader.read(null, new BinaryDecoder(in));
    assertEquals(a, decoded);
    decoded = reader.read(null, new BinaryDecoder(in));
    assertEquals(b, decoded);
  }

  public static class SampleRecord {
    public int x = 1;
    private int y = 2;

    public int hashCode() {
      return x + y;
    }

    public boolean equals(Object obj) {
      if (this == obj)
        return true;
      if (obj == null)
        return false;
      if (getClass() != obj.getClass())
        return false;
      final SampleRecord other = (SampleRecord)obj;
      if (x != other.x)
        return false;
      if (y != other.y)
        return false;
      return true;
    }
    
    public static class AnotherSampleRecord {
      private Integer a = null;
      private SampleRecord s = null;

      public AnotherSampleRecord() {
      }

      AnotherSampleRecord(Integer a) {
        this.a = a;
        this.s = new SampleRecord();
      }

      public int hashCode() {
        int hash = (a != null ? a.hashCode() : 0);
        hash += (s != null ? s.hashCode() : 0);
        return hash;
      }

      public boolean equals(Object other) {
        if (other instanceof AnotherSampleRecord) {
          AnotherSampleRecord o = (AnotherSampleRecord) other;
          if ( (this.a == null && o.a != null) ||
               (this.a != null && !this.a.equals(o.a)) ||
               (this.s == null && o.s != null) ||
               (this.s != null && !this.s.equals(o.s)) ) {
            return false;
          }
          return true;
        } else {
          return false;
        }
      }
    }
  }

  public static class X { int i; }
  public static class B1 { X x; }
  public static class B2 { X x; }
  public static class A { B1 b1; B2 b2; }
  public static interface C { void foo(A a); }

  @Test
  public void testForwardReference() {
    ReflectData data = ReflectData.get();
    Protocol reflected = data.getProtocol(C.class);
    Protocol reparsed = Protocol.parse(reflected.toString());
    assertEquals(reflected, reparsed);
    assert(reparsed.getTypes().contains(data.getSchema(A.class)));
    assert(reparsed.getTypes().contains(data.getSchema(B1.class)));
    assert(reparsed.getTypes().contains(data.getSchema(B2.class)));
    assert(reparsed.getTypes().contains(data.getSchema(X.class)));
  }

}

File: lang/csharp/src/apache/test/Reflect/TestReflect.cs
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections;
using System.IO;
using NUnit.Framework;
using Avro.IO;
using Avro.Reflect;
using System.Collections.Generic;
using System;
using Avro.Specific;
using System.Linq;
using System.Diagnostics.CodeAnalysis;

namespace Avro.Test
{
    [TestFixture]
    class TestReflect
    {
        public class IDictionaryTestClass : IDictionary<string, int>
        {
            Dictionary<string, int> _pairs = new Dictionary<string, int>();
            public int this[string key] { get => _pairs[key]; set => _pairs[key] = value; }

            public ICollection<string> Keys => _pairs.Keys;

            public ICollection<int> Values => _pairs.Values;

            public int Count => _pairs.Count;

            public bool IsReadOnly => throw new NotImplementedException("readonly");

            public void Add(string key, int value) => _pairs.Add(key, value);
            public void Add(KeyValuePair<string, int> item) => throw new NotImplementedException("add_kvp");
            public void Clear() => throw new NotImplementedException("clear");
            public bool Contains(KeyValuePair<string, int> item) => throw new NotImplementedException("contains");
            public bool ContainsKey(string key) => throw new NotImplementedException("containskey");
            public void CopyTo(KeyValuePair<string, int>[] array, int arrayIndex) => throw new NotImplementedException("copyto");
            public IEnumerator<KeyValuePair<string, int>> GetEnumerator() => _pairs.GetEnumerator();
            public bool Remove(string key) => throw new NotImplementedException("remove");
            public bool Remove(KeyValuePair<string, int> item) => throw new NotImplementedException("remove_kvp");
            public bool TryGetValue(string key, out int value) => throw new NotImplementedException("trygetvalue");
            IEnumerator IEnumerable.GetEnumerator() => _pairs.GetEnumerator();
            public static implicit operator IDictionaryTestClass(Dictionary<string,int> instance)
            {
                IDictionaryTestClass result = new IDictionaryTestClass();
                foreach (var item in instance)
                {
                    result.Add(item.Key, item.Value);
                }
                return result;
            }
        }

        public class TestClassNestedInheritance : IDictionaryTestClass {}

        public class TestClassWithPropertyTestingInheritance
        {
            public TestClassNestedInheritance prop { get; set; }
        }

        public class DictionaryTestClass
        {
            public Dictionary<int, int> p { get; set; }
            public NestedTestClass ntc { get; set; }


            public class NestedTestClass
            {
                public int NestedTestClassInt { get; set; }
            }
        }

        public class DictionaryTestClass2
        {
            public Dictionary<int, string> p { get; set; }

        }
        class TestMapConverter : IAvroFieldConverter
        {
            public object FromAvroType(object o, Schema s) =>
                ((Dictionary<string, int>)o).ToDictionary(x => int.Parse(x.Key), y=>y.Value);
                
            
            public Type GetAvroType() => typeof(IDictionary<string, int>);
            public Type GetPropertyType() => typeof(Dictionary<int, int>);
            public object ToAvroType(object o, Schema s) =>
                ((Dictionary<int, int>)o).ToDictionary(x => x.Key.ToString(), y => y.Value);
        }

        class TestConverterForNestedInheritance : IAvroFieldConverter
        {
            public object FromAvroType(object o, Schema s)
            {
                TestClassNestedInheritance result = new TestClassNestedInheritance();
                foreach(var kvp in (Dictionary<string, int>)o)
                {
                    result.Add(kvp.Key, kvp.Value);
                }
                return result;
            }


            public Type GetAvroType() => typeof(IDictionary<string, int>);
            public Type GetPropertyType() => typeof(TestClassNestedInheritance);
            public object ToAvroType(object o, Schema s) =>
                ((IDictionary<string, int>)o).ToDictionary(x => x.Key.ToString(), y => y.Value);
        }


        [TestCase]
        public void TestMapWithConverterSucceeds()
        {
            ClassCache.AddDefaultConverter(new TestMapConverter());
            
            var schemaJson = "{\"fields\":[{\"name\":\"ntc\",\"type\":{\"type\":\"record\",\"name\":\"NestedTestClass\",\"fields\":[{\"name\":\"NestedTestClassInt\",\"type\":\"int\"}]}},{\"type\":{\"values\":\"int\",\"type\":\"map\"},\"name\":\"p\"}],\"type\":\"record\",\"name\":\"DictionaryTestClass\",\"namespace\":\"Avro.Test.TestReflect\\u002B\"}";
            var schema = Schema.Parse(schemaJson);
            DictionaryTestClass expected = new DictionaryTestClass() { p = new Dictionary<int, int>() { { 1, 1 }, { 2, 4 }, { 3, 5 } } , ntc = new DictionaryTestClass.NestedTestClass() { NestedTestClassInt = 1 } };

            using Stream stream =  serialize(schema, expected);
            var avroReader = new ReflectReader<DictionaryTestClass>(schema, schema);    
            stream.Position = 0;
            var actual = avroReader.Read(null, new BinaryDecoder(stream));
           
            CollectionAssert.AreEquivalent(expected.p, actual.p);
        }

        [TestCase]
        public void TestMapWithoutConverterFails()
        {
            var schemaJson = "{\"fields\":[{\"type\":{\"values\":\"string\",\"type\":\"map\"},\"name\":\"p\"}],\"type\":\"record\",\"name\":\"DictionaryTestClass2\",\"namespace\":\"Avro.Test.TestReflect\\u002B\"}";
            var schema = Schema.Parse(schemaJson);
            var ex = Assert.Throws<AvroException>(() => new ReflectWriter<DictionaryTestClass2>(schema));
            var ex2 = Assert.Throws<AvroException>(() => new ReflectReader<DictionaryTestClass2>(schema, schema));
            Assert.AreEqual("Property p in object Avro.Test.TestReflect+DictionaryTestClass2 isn't compatible with Avro schema type Map", ex.Message);
            Assert.AreEqual("Property p in object Avro.Test.TestReflect+DictionaryTestClass2 isn't compatible with Avro schema type Map", ex2.Message);
        }

        public class DictionaryTestClass3
        {
            public Hashtable p { get; set; }
        }

        [TestCase]
        public void TestHashmapFailsBecauseItDoesntImplmentIDictionaryTKeyTValue()
        {
            var schemaJson = "{\"fields\":[{\"type\":{\"values\":\"string\",\"type\":\"map\"},\"name\":\"p\"}],\"type\":\"record\",\"name\":\"DictionaryTestClass3\",\"namespace\":\"Avro.Test.TestReflect\\u002B\"}";
            var schema = Schema.Parse(schemaJson);            
            var ex = Assert.Throws<AvroException>(() => new ReflectWriter<DictionaryTestClass3>(schema));
            Assert.AreEqual("Property p in object Avro.Test.TestReflect+DictionaryTestClass3 isn't compatible with Avro schema type Map", ex.Message);

        }


        [TestCase]
        public void TestConverterForNestedInheritanceSucceeds()
        {
            ClassCache.AddDefaultConverter(new TestConverterForNestedInheritance());

            var schemaJson = "{\"fields\":[{\"type\":{\"values\":\"int\",\"type\":\"map\"},\"name\":\"prop\"}],\"type\":\"record\",\"name\":\"TestClassWithTestClassA\",\"namespace\":\"Avro.Test.TestReflect\\u002B\"}";
            var schema = Schema.Parse(schemaJson);


            TestClassWithPropertyTestingInheritance expected = new TestClassWithPropertyTestingInheritance() { prop = new TestClassNestedInheritance() };
            expected.prop["1"] = 1;

            using Stream stream = serialize(schema, expected);
            var avroReader = new ReflectReader<TestClassWithPropertyTestingInheritance>(schema, schema);
            stream.Position = 0;
            var actual = avroReader.Read(null, new BinaryDecoder(stream));

            CollectionAssert.AreEquivalent(expected.prop, actual.prop);   

        }

        enum EnumResolutionEnum
        {
            THIRD,
            FIRST,
            SECOND
        }

        class EnumResolutionRecord
        {
            public EnumResolutionEnum enumType { get; set; }
        }

        [TestCase]
        public void TestEnumResolution()
        {
            Schema writerSchema = Schema.Parse("{\"type\":\"record\",\"name\":\"EnumRecord\",\"namespace\":\"Avro.Test\"," +
                                        "\"fields\":[{\"name\":\"enumType\",\"type\": { \"type\": \"enum\", \"name\": \"EnumType\", \"symbols\": [\"FIRST\", \"SECOND\"]} }]}");

            var testRecord = new EnumResolutionRecord();

            Schema readerSchema = Schema.Parse("{\"type\":\"record\",\"name\":\"EnumRecord\",\"namespace\":\"Avro.Test\"," +
                                        "\"fields\":[{\"name\":\"enumType\",\"type\": { \"type\": \"enum\", \"name\":" +
                                        " \"EnumType\", \"symbols\": [\"THIRD\", \"FIRST\", \"SECOND\"]} }]}");;
            testRecord.enumType = EnumResolutionEnum.SECOND;

            // serialize
            var stream = serialize(writerSchema, testRecord);

            // deserialize
            var rec2 = deserialize<EnumResolutionRecord>(stream, writerSchema, readerSchema);
            Assert.AreEqual( EnumResolutionEnum.SECOND, rec2.enumType );
        }

        private static S deserialize<S>(Stream ms, Schema ws, Schema rs) where S : class
        {
            long initialPos = ms.Position;
            var r = new ReflectReader<S>(ws, rs);
            Decoder d = new BinaryDecoder(ms);
            S output = r.Read(null, d);
            Assert.AreEqual(ms.Length, ms.Position); // Ensure we have read everything.
            checkAlternateDeserializers(output, ms, initialPos, ws, rs);
            return output;
        }

        private static void checkAlternateDeserializers<S>(S expected, Stream input, long startPos, Schema ws, Schema rs) where S : class
        {
            input.Position = startPos;
            var reader = new ReflectReader<S>(ws, rs);
            Decoder d = new BinaryDecoder(input);
            S output = reader.Read(null, d);
            Assert.AreEqual(input.Length, input.Position); // Ensure we have read everything.
            AssertReflectRecordEqual(rs, expected, ws, output, reader.Reader.ClassCache);
        }

        private static Stream serialize<T>(Schema ws, T actual)
        {
            var ms = new MemoryStream();
            Encoder e = new BinaryEncoder(ms);
            var w = new ReflectWriter<T>(ws);
            w.Write(actual, e);
            ms.Flush();
            ms.Position = 0;
            checkAlternateSerializers(ms.ToArray(), actual, ws);
            return ms;
        }

        private static void checkAlternateSerializers<T>(byte[] expected, T value, Schema ws)
        {
            var ms = new MemoryStream();
            var writer = new ReflectWriter<T>(ws);
            var e = new BinaryEncoder(ms);
            writer.Write(value, e);
            var output = ms.ToArray();

            Assert.AreEqual(expected.Length, output.Length);
            Assert.True(expected.SequenceEqual(output));
        }

        private static void AssertReflectRecordEqual(Schema schema1, object rec1, Schema schema2, object rec2, ClassCache cache)
        {
            var recordSchema = (RecordSchema) schema1;
            foreach (var f in recordSchema.Fields)
            {
                var rec1Val = cache.GetClass(recordSchema).GetValue(rec1, f);
                var rec2Val = cache.GetClass(recordSchema).GetValue(rec2, f);
                if (rec1Val.GetType().IsClass)
                {
                    AssertReflectRecordEqual(f.Schema, rec1Val, f.Schema, rec2Val, cache);
                }
                else if (rec1Val is IList) //I don't see when this or the last block would be evaluated.  Isn't anything that is IList or IDictionary going to be a class?
                {
                    var schema1List = f.Schema as ArraySchema;
                    var rec1List = (IList) rec1Val;
                    if( rec1List.Count > 0 )
                    {
                        var rec2List = (IList) rec2Val;
                        Assert.AreEqual(rec1List.Count, rec2List.Count);
                        for (int j = 0; j < rec1List.Count; j++)
                        {
                            AssertReflectRecordEqual(schema1List.ItemSchema, rec1List[j], schema1List.ItemSchema, rec2List[j], cache);
                        }
                    }
                    else
                    {
                        Assert.AreEqual(rec1Val, rec2Val);
                    }
                }
                else if (rec1Val is IDictionary)
                {
                    var schema1Map = f.Schema as MapSchema;
                    var rec1Dict = (IDictionary) rec1Val;
                    var rec2Dict = (IDictionary) rec2Val;
                    Assert.AreEqual(rec2Dict.Count, rec2Dict.Count);
                    foreach (var key in rec1Dict.Keys)
                    {
                        var val1 = rec1Dict[key];
                        var val2 = rec2Dict[key];
                        if (f.Schema is RecordSchema)
                        {
                            AssertReflectRecordEqual(f.Schema as RecordSchema, val1, f.Schema as RecordSchema, val2, cache);
                        }
                        else
                        {
                            Assert.AreEqual(val1, val2);
                        }
                    }
                }
                else
                {
                    Assert.AreEqual(rec1Val, rec2Val);
                }
            }
        }
    }
}
